### PR TITLE
Cypress Automated Test for Feedback Modal and Additional Fixes for Various Tests

### LINKED ISSUE #1513: E2E Cypress test - Test "Login fail alert appears when incorrect email & password are entered and submitted by user" fails
Env: Dev

Test "Login fail alert appears when incorrect email & password are entered and submitted by user" fails when it runs multiple times.
Expected error message: Login failed, reason: Wrong email or password. 
Actual error message: Login failed. Maximum number of sign in attempts exceeded.

```
 User signs into account
cypress-e2e          |          Login fail alert appears when incorrect email & password are entered and submitted by user:
cypress-e2e          |      AssertionError: Timed out retrying: Expected to find content: 'Login failed, reason: Wrong email or password.' within the element: <span.ant-alert-message> but never did.
cypress-e2e          |       at Context.eval (http://client:3000/__cypress/tests?p=cypress/integration/e2eTests/signin/signIn.spec.js:256:22)
```

FightPandemics Sign In Page -- User signs into account -- Login fail alert appears when incorrect email   password are entered and submitted by user (failed)

### PR DESCRIPTION
**Pull request by jessica-lau**
### What does this pull request aim to achieve? 💯

1. Created automated tests for Feedback Modal screens. 

2. Updated code for the location input in the Request Help questionnaire test.

3. Updated code to include {force: true} within .click() in Footer test to correct occasional error arising due to center of element hidden from view.

4. Updated code in Sign In test to fix Issue: [1513](https://github.com/FightPandemics/FightPandemics/issues/1513)

_Please be concise and link any related issue(s):_

- [x] I have checked that no one else is working on similar changes.
- [x] I have read the latest [**README**](README.md) and understand the development workflow.
- [x] The name of this branch is: **`feature/<branch_name>`**, or **`hotfix/<branch_name>`** if this is a hotfix. The branch is created in a cloned version of this repo, **not a forked repo**.
- [x] This branch is rebased/merged with the **latest staging** (or the **latest production** if it is a hotfix).
- [x] There are no merge conflicts.
- [x] There are no console warnings and errors.
- [x] On the right hand side of this PR, I have linked the appropriate issue(s) under "Issues" and added a project progress under "Projects".

### CODE REVIEW

File: cypresse2e/cypress/integration/e2eTests/feedbackModal/feedbackModal.spec.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            });
+            it('Choosing rating 4', () => {
+                feedback.getFeedbackScale().find('div').eq(3).click();
+            });
+            it('Choosing rating 5', () => {
+                feedback.getFeedbackScale().find('div').eq(4).click();
+            });
+
+        
+            // it.skip('Feedback scale numbers are visible and clickable', () => {
```
Comment (galeza) [inline cypresse2e/cypress/integration/e2eTests/feedbackModal/feedbackModal.spec.js]: Hi @jessica-lau 
Are we going to use this commented out code? Or we can remove it?

**Review by galeza — CHANGES_REQUESTED** [review]

**Comment by galeza** [comment]
@jessica-lau 
Should I add a task for next Feedback modal screens: Thank you for being an early user of FightPandemics! and We are almost done!? Let me know.

**Comment by jessica-lau** [comment]
@galeza I'm actually still working on writing up this test for the feedback modal, so it's not ready for review yet. I have two more feedback screens to write the test code for (the ones you mentioned above) and hope to finish it over the weekend. I just converted this PR to a draft instead. Sorry for the confusion! I'll assign this for review once it's ready. Thank you! :)

File: cypresse2e/cypress/integration/e2eTests/feedbackModal/feedbackModal.spec.js (hunk 1 above)
Comment (jessica-lau) [reply cypresse2e/cypress/integration/e2eTests/feedbackModal/feedbackModal.spec.js]: Yes, I'll remove it when I'm finish with writing the test. 👍

**Comment by galeza** [comment]
Hi @jessica-lau 
I run this branch. Footer spec works fine.
Only this test fails:
```
cypress-e2e          |   1) FightPandemics Sign In Page
cypress-e2e          |        User signs into account
cypress-e2e          |          Login fail alert appears when incorrect email & password are entered and submitted by user:
cypress-e2e          |      AssertionError: Timed out retrying: Expected to find content: 'Login failed, reason: Wrong email or password.' within the element: <span.ant-alert-message> but never did.
cypress-e2e          |       at Context.eval (http://client:3000/__cypress/tests?p=cypress/integration/e2eTests/signin/signIn.spec.js:261:22)
```

**Comment by jessica-lau** [comment]
Hey @galeza for the 'Sign In' page I changed the code to include generating a random email in order to trigger the "Login failed, reason: Wrong email or password", like how you did for the mocha backend test. Can you check to see if it works for you now?

**Comment by galeza** [comment]
Hi @jessica-lau 

I tested this branch with randomStringGenerator. All SignIn tests pass.
I would only move randomString code to the commands.js. It's more Cypress like. Code that can be share can be put there. It would look like this:
```
Cypress.Commands.add('randomString', (length) => { 
    var result           = '';
    var characters       = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    var charactersLength = characters.length;
    for ( var i = 0; i < length; i++ ) {
       result += characters.charAt(Math.floor(Math.random() * charactersLength));
    }
    return result;
});
```
And in the test cy.randomString(7);

Only if you have cypress runner open, you will need to close it and reopen before starting tests.

**Comment by jessica-lau** [comment]
@galeza After moving the randomString to the commands.js file using the following, the test fails:

```
Cypress.Commands.add('randomString', (length) => { 
    var result           = '';
    var characters       = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    var charactersLength = characters.length;
    for ( var i = 0; i < length; i++ ) {
       result += characters.charAt(Math.floor(Math.random() * charactersLength));
    }
    return result;
});
```

I tried two different ways of formatting the code in the Sign In test but encountered errors stating that I'm passing an object or it requires a DOM element. It seems that the code doesn't execute as expected when formatted as a command. Am I missing something?

1. First formatting

```
it('Login fail alert appears when incorrect email & password are entered and submitted by user', () => {
            var email = cy.randomString(5);
            signIn.getEmailField().type(email + '.' + email + '@' + email + '.com');
            signIn.getPasswordField().type('WrongPW!wpw3hi');
            signIn.getSignInButton().click();
            var loginFailAlert = signIn.getLoginFailAlert();
            loginFailAlert.should('be.visible');
            loginFailAlert.contains('Login failed, reason: Wrong email or password.');
        });
```

2. Second formatting

```
it('Login fail alert appears when incorrect email & password are entered and submitted by user', () => {
            signIn.getEmailField().type(cy.randomString(8) + '.' + cy.randomString(8) + '@' + cy.randomString(5) + '.com');
            signIn.getPasswordField().type('WrongPW!wpw3hi');
            signIn.getSignInButton().click();
            var loginFailAlert = signIn.getLoginFailAlert();
            loginFailAlert.should('be.visible');
            loginFailAlert.contains('Login failed, reason: Wrong email or password.');
        });
```

Cypress error:

image

**Comment by galeza** [comment]
Hi @jessica-lau 
Try this:
commands.js
```
export const randomString = (length) => {
    var result = '';
    var characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    var charactersLength = characters.length;
    for ( var i = 0; i < length; i++ ) {
       result += characters.charAt(Math.floor(Math.random() * charactersLength));
    }
    return result;
  };

Cypress.Commands.add('generateRandomEmail', () => { 
    return randomString(8) + '.' + randomString(8) + '@' + randomString(5) + '.com';
});
```
Then signIn.spec.js:

```
        it('Login fail alert appears when incorrect email & password are entered and submitted by user', () => {
            cy.generateRandomEmail().then((email) => { 
                signIn.getEmailField().type(email);
            });   
                 
            signIn.getPasswordField().type('WrongPW!wpw3hi');
            signIn.getSignInButton().click();
            var loginFailAlert = signIn.getLoginFailAlert();
            loginFailAlert.should('be.visible');
            loginFailAlert.contains('Login failed, reason: Wrong email or password.');
        });
```

**Comment by jessica-lau** [comment]
@galeza this worked. Thanks for all your help!

**Review by galeza — APPROVED** [review]

**Comment by sheeba1** [comment]
I ran the tests locally. 2 failed for me (could be delays) - 1. Got this error for feedback.



image



2. Request help


image

**Comment by jessica-lau** [comment]
@sheeba1 the error for the feedback is most likely due to a delay. Can you please try re-running? It passes with no issues for me. 

Regarding the error in the request help, the "Next" arrow on question 2 has been removed recently since question 3 has been completely deleted. Refer to PR#1525 for details. I just updated the code with the change to reflect this.

**Comment by sheeba1** [comment]
Thanks @jessica-lau  for the cod eupdate , it worked fine for 'Request Help'. 


Still getting the first error for, feedback.

It should be ok , as it works fine for @jessica-lau and @galeza

**Comment by galeza** [comment]
Hi,

I run this branch several times and in docker-compose it was ok. Sometimes I see ResizeObserver loop limit exceeded error so I added a code in the /support/index.js file preventing this error. It can happen with Cypress version 4.10.0.
Apart that everything works fine. Maybe errors were caused by runner?

Screen Shot 2020-09-18 at 12 27 07

**Review by sheeba1 — APPROVED** [review]

**Review by idjevm — APPROVED** [review]
LGTM! Thanks @jessica-lau

**Comment by idjevm** [comment]
Let me know if you want this merged please :)

**Comment by jessica-lau** [comment]
@idjevm yes please merge this to staging. Thank you!

**Pull request merged**